multiboot2/src/memory_map.rs

```diff
@@ -140,7 +140,7 @@ impl EFIMemoryMapTag {
         let start_area = (&self.first_desc) as *const EFIMemoryDesc;
         EFIMemoryAreaIter {
             current_area: start_area as u64,
-            last_area: (self_ptr as u64 + self.size as u64),
+            last_area: (self_ptr as u64 + (self.size - self.desc_size) as u64),
```

Comment:
I think that the root problem is that `size_of::<EFIMemoryMapTag>()` contains already the size of the first entry. I think a nicer solution would be to type the `EFIMemoryMapTag` as

```rust
pub struct EFIMemoryMapTag {
    typ: TagType,
    size: u32,
    desc_size: u32,
    desc_version: u32,
    first_desc: [EFIMemoryDesc; 0],
}
```

instead. Depending on your skill and time you'd like to invest, feel free to look into it. Otherwise, I can take care of it.

Reply:
Thanks for the suggestion. I'm a bit new to Rust, but the zero length arrays seem much more elegant.

I've updated both the E820 and EFI code to use them.

Reply:
looks good!